README.md
# LivingLogic.Python.pycoco

Python code coverage

Walter Doerwald


NEWS.rst
Changes in 0.8 (released ??/??/2013)
------------------------------------

*	The Python source code is checked out via ``hg`` now.

*	Updated to use Python 3.3 and XIST 4.10.


Changes in 0.7.2 (released 09/08/2010)
--------------------------------------

*	Added the missing pixel GIF ``spc.gif``.


Changes in 0.7.1 (released 08/11/2010)
--------------------------------------

*	Fixed a bug in the :meth:`download` method.


Changes in 0.7 (released 08/10/2010)
------------------------------------

*	The URL of the Python tarball can now be specified on the command line.

*	Generating coverage info is now done via calling ``make coverage``.


Changes in 0.6 (released 05/13/2009)
------------------------------------

*	Generating the HTML files is now done via UL4 templates (which takes 40
	seconds instead of 30 minutes).


Changes in 0.5 (released 03/18/2008)
------------------------------------

*	``pycoco`` now uses the ``-coverage`` option for compiling and linking
	(This requires gcc 4).

*	Two logfiles are generated now: A log of the build process and a log of
	running the tests. Both files are accessible from the generated start page.

*	The following tests are now run again: ``urlfetch``, ``largefile``,
	``network`` and ``decimal``


Changes in 0.4 (released 02/25/2008)
------------------------------------

*	Removed XML version of the documentation.

*	Only the default set of tests gets run now.

*	Fixed a bug in page generation.


Changes in 0.3 (released 01/07/2008)
------------------------------------

*	Changed docstring format.

*	Updated to use XIST 3.0.


Changes in 0.2 (released 10/19/2006)
------------------------------------

*	The list of files can now be sorted by clicking on the appropriate buttons
	in the table headers.


Changes in 0.1.1 (released 10/02/2006)
--------------------------------------

*	Added a missing import to the XIST namespace.


Changes in 0.1 (released 10/02/2006)
------------------------------------

*	Initial release.


src/pycoco/xmlns.py
# -*- coding: utf-8 -*-

"""
This module is an XIST namespace used for formatting the HTML coverage report.
"""

import datetime

from ll.xist import xsc
from ll.xist.ns import xml, html, meta, htmlspecials, ul4


xmlns = "http://xmlns.python.org/coverage"


class page(xsc.Element):
	xmlns = xmlns

	def convert(self, converter):
		e = xsc.Frag(
			xml.XML(), "\n",
			html.DocTypeXHTML10transitional(), "\n",
			html.html(
				html.head(
					meta.contenttype(),
					html.title(
						"Python code coverage",
						ul4.if_("filename"),
							": ", ul4.printx("filename"),
						ul4.else_(),
							" (", ul4.print_("format(timestamp, '%Y-%m-%d')"), ")",
						ul4.end("if"),
					),
					meta.stylesheet(href="/coverage.css"),
					meta.stylesheet(href="/coverage_sortfilelist.css"),
					htmlspecials.javascript(src="/coverage.js"),
				),
				html.body(
					html.div(
						html.div(
							html.a(
								htmlspecials.autoimg(src="http://www.python.org/images/python-logo.gif", alt="Python", border=0),
								href="http://www.python.org/",
							),
							class_="logo",
						),
						html.div(
							ul4.for_("(i, item) in enumerate(crumbs)"),
								html.span(
									html.span(
										ul4.if_("i"),
											">",
										ul4.else_(),
											"\xbb",
										ul4.end("if"),
										class_="bullet",
									),
									ul4.if_("item.href"),
										html.a(ul4.printx("item.title"), href=ul4.printx("item.href")),
									ul4.else_(),
										html.span(ul4.printx("item.title"), class_="here"),
									ul4.end("if"),
								ul4.end("for"),
								class_="crumb",
							),
							class_="crumbs",
						),
						class_="header",
					),
					html.div(
						self.content,
						class_="content",
					),
					onload=ul4.attr_if(ul4.printx("onload"), cond="onload"),
				),
			),
		)
		return e.convert(converter)


class filelist(xsc.Element):
	xmlns = xmlns

	def convert(self, converter):
		e = xsc.Frag(
			html.h1("Python code coverage"),
			html.p("Generated at ", ul4.printx("format(now, '%Y-%m-%d %H:%M:%S')"), class_="note"),
			html.p("Last commit at ", ul4.printx("format(timestamp, '%Y-%m-%d %H:%M:%S')"), " by ", ul4.printx("author"), class_="note"),
			html.p("Changeset identification hash ", ul4.printx("changesetid"), class_="note"),
			html.p("Local revision number ", ul4.printx("revision"), class_="note"),
			html.p(html.a("Build log", href="buildlog.txt"), " ",html.a("Test log", href="testlog.txt"), class_="note"),
			htmlspecials.plaintable(
				html.thead(
					html.tr(
						html.th("Filename", id="filename"),
						html.th("# lines", id="nroflines"),
						html.th("# coverable lines", id="coverablelines"),
						html.th("# covered lines", id="coveredlines"),
						html.th("coverage", id="coverage"),
						html.th("distribution", id="distibution"),
					),
				),
				html.tbody(
					ul4.for_("file in files"),
						html.tr(
							html.th(
								html.a(
									ul4.printx("file.name"),
									href=(ul4.printx("file.name"), ".html"),
								),
								class_="filename",
							),
							html.td(
								ul4.printx("file.lines"),
								class_="nroflines",
							),
							html.td(
								ul4.printx("file.coverablelines"),
								class_="coverablelines",
							),
							html.td(
								ul4.printx("file.coveredlines"),
								class_="coveredlines",
							),
							html.td(
								ul4.if_("file.coverablelines"),
									ul4.printx("format((100.*file.coveredlines)/file.coverablelines, '.2f')"),
									"%",
								ul4.else_(),
									"n/a",
								ul4.end("if"),
								class_=(
									"coverage",
									ul4.if_("not file.coverablelines"),
										" disabled",
									ul4.end("if"),
								),
							),
							html.td(
								ul4.code("totalwidth = 100"),
								ul4.if_("file.coverablelines"),
									ul4.if_("file.coverablelines < file.lines"),
										ul4.code("width = int(1.*(file.lines-file.coverablelines)/file.lines*100)"),
										htmlspecials.pixel(src="/spc.gif", width=ul4.printx("width"), height=8, style="background-color: #ccc;"),
										ul4.code("totalwidth -= width"),
									ul4.end("if"),
									ul4.if_("file.coveredlines < file.coverablelines"),
										ul4.code("width = int(1.*(file.coverablelines-file.coveredlines)/file.lines*100)"),
										htmlspecials.pixel(src="/spc.gif", width=ul4.printx("width"), height=8, style="background-color: #f00;"),
										ul4.code("totalwidth -= width"),
									ul4.end("if"),
									ul4.if_("totalwidth"),
										htmlspecials.pixel(src="/spc.gif", width=ul4.printx("totalwidth"), height=8, style="background-color: #0c0;"),
									ul4.end("if"),
								ul4.else_(),
									htmlspecials.pixel(src="/spc.gif", width=ul4.printx("totalwidth"), height=8, style="background-color: #000;"),
								ul4.end("if"),
								class_="dist",
							),
							class_="files",
						),
					ul4.end("for"),
					id="files",
				),
				class_="files",
			)
		)
		return e.convert(converter)


class filecontent(xsc.Element):
	xmlns = xmlns

	def convert(self, converter):
		e = xsc.Frag(
			html.h1("Python code coverage for ", ul4.printx("filename")),
			htmlspecials.plaintable(
				html.thead(
					html.tr(
						html.th("#"),
						html.th("count"),
						html.th("content"),
					),
				),
				html.tbody(
					ul4.for_("(i, line) in enumerate(lines)"),
						html.tr(
							html.th(ul4.print_("i+1")),
							html.td(
								ul4.if_("not isnone(line.count) and line.count >= 0"),
									ul4.printx("line.count"),
								ul4.else_(),
									"n/a",
								ul4.end("if"),
								class_="count",
							),
							html.td(ul4.printx("line.content"), class_="line"),
							class_=ul4.attr_if(
								ul4.if_("isnone(line.count) or line.count < 0"),
									"uncoverable",
								ul4.elif_("not line.count"),
									"uncovered",
								ul4.end("if"),
								cond="isnone(line.count) or line.count <= 0",
							),
						),
					ul4.end("for"),
				),
				class_="file",
			)
		)
		return e.convert(converter)


src/pycoco/__init__.py
# -*- coding: utf-8 -*-


from __future__ import with_statement

import sys, os, datetime, urllib, optparse, contextlib

from ll import sisyphus, url, ul4c

from pycoco import xmlns


class File(object):
	def __init__(self, name):
		self.name = name
		self.lines = [] # list of lines with tuples (# of executions, line)

	def __repr__(self):
		return "<File name=%r at 0x%x>" % (self.name, id(self))


class Python_GenerateCodeCoverage(sisyphus.Job):
	def __init__(self, outputdir):
		sisyphus.Job.__init__(self, 2*60*60, name="Python_GenerateCodeCoverage", raiseerrors=1)
		self.url = "http://svn.python.org/snapshots/python.tar.bz2"
		self.tarfile = "python.tar.bz2"
		self.outputdir = url.Dir(outputdir)

		self.configurecmd = "./configure --enable-unicode=ucs4 --with-pydebug"

		self.gcovcmd = os.environ.get("COV", "gcov")
		self.makefile = "python/Makefile"

		self.buildlog = [] # the output of configuring and building Python
		self.testlog = [] # the output of running the test suite

	def cmd(self, cmd):
		self.logProgress(">>> %s" % cmd)
		pipe = os.popen(cmd + " 2>&1", "rb", 1)
		lines = []
		for line in pipe:
			self.logProgress("... " + line)
			lines.append(line)
		return lines

	def files(self, base):
		self.logProgress("### finding files")
		allfiles = []
		for (root, dirs, files) in os.walk(base):
			for file in files:
				if file.endswith(".py") or file.endswith(".c"):
					allfiles.append(File(os.path.join(root, file)))
		self.logProgress("### found %d files" % len(allfiles))
		return allfiles

	def download(self):
		self.logProgress("### downloading %s to %s" % (self.url, self.tarfile))
		urllib.urlretrieve(self.url, self.tarfile)

	def unpack(self):
		self.logProgress("### unpacking %s" % self.tarfile)
		self.cmd("tar xvjf %s" % self.tarfile)
		lines = list(open("python/.timestamp", "r"))
		self.timestamp = datetime.datetime.fromtimestamp(int(lines[0]))
		self.revision = lines[2]

	def configure(self):
		self.logProgress("### configuring")
		lines = self.cmd("cd python; %s" % self.configurecmd)
		self.buildlog.extend(lines)

	def make(self):
		self.logProgress("### running make")
		self.buildlog.extend(self.cmd("cd python && make coverage"))

	def test(self):
		self.logProgress("### running test")
		lines = self.cmd("cd python && ./python Lib/test/regrtest.py -T -N -uurlfetch,largefile,network,decimal")
		self.testlog.extend(lines)

	def cleanup(self):
		self.logProgress("### cleaning up files from previous run")
		self.cmd("rm -rf python")
		self.cmd("rm %s" % self.tarfile)

	def coveruncovered(self, file):
		self.logProgress("### faking coverage info for uncovered file %r" % file.name)
		file.lines = [(None, line.rstrip("\n")) for line in open(file.name, "r")]

	def coverpy(self, file):
		coverfilename = os.path.splitext(file.name)[0] + ".cover"
		self.logProgress("### fetching coverage info for Python file %r from %r" % (file.name, coverfilename))
		try:
			f = open(coverfilename, "r")
		except IOError, exc:
			self.logError(exc)
			self.coveruncovered(file)
		else:
			for line in f:
				line = line.rstrip("\n")
				prefix, line = line[:7], line[7:]
				prefix = prefix.strip()
				if prefix == "." or prefix == "":
					file.lines.append((-1, line))
				elif prefix == ">"*len(prefix):
					file.lines.append((0, line))
				else:
					file.lines.append((int(prefix.rstrip(":")), line))
			f.close()

	def coverc(self, file):
		self.logProgress("### fetching coverage info for C file %r" % file.name)
		dirname = os.path.dirname(file.name).split("/", 1)[-1]
		basename = os.path.basename(file.name)
		self.cmd("cd python && %s %s -o %s" % (self.gcovcmd, basename, dirname))
		try:
			f = open("python/%s.gcov" % basename, "r")
		except IOError, exc:
			self.logError(exc)
			self.coveruncovered(file)
		else:
			for line in f:
				line = line.rstrip("\n")
				if line.count(":") < 2:
					continue
				(count, lineno, line) = line.split(":", 2)
				count = count.strip()
				lineno = lineno.strip()
				if lineno == "0": # just the header
					continue
				if count == "-": # not executable
					file.lines.append((-1, line))
				elif count == "#####": # not executed
					file.lines.append((0, line))
				else:
					file.lines.append((int(count), line))
			f.close()

	def makehtml(self, files):
		# Generate main page
		self.logProgress("### generating index page")
		template = ul4c.compile(xmlns.page(xmlns.filelist(), onload="files_prepare()").conv().string())
		s = template.renders(
			filename=None,
			now=datetime.datetime.now(),
			timestamp=self.timestamp,
			revision=self.revision,
			crumbs=[
				dict(title="Core Development", href="http://www.python.org/dev/"),
				dict(title="Code coverage", href=None),
			],
			files=[
				dict(
					name=file.name.split("/", 1)[-1],
					lines=len(file.lines),
					coverablelines=sum(line[0]>=0 for line in file.lines),
					coveredlines=sum(line[0]>0 for line in file.lines),
				) for file in files
			],
		)
		u = self.outputdir/"index.html"
		with contextlib.closing(u.openwrite()) as f:
			f.write(s.encode("utf-8"))

		# Generate page for each source file
		template = ul4c.compile(xmlns.page(xmlns.filecontent()).conv().string())
		for (i, file) in enumerate(files):
			filename = file.name.split("/", 1)[-1]
			self.logProgress("### generating HTML %d/%d for %s" % (i+1, len(files), filename))
			s = template.renders(
				filename=filename,
				crumbs=[
					dict(title="Core Development", href="http://www.python.org/dev/"),
					dict(title="Code coverage", href="/index.html"),
					dict(title=filename, href=None),
				],
				lines=(
					dict(count=count, content=content.decode("latin-1").expandtabs(8)) for (count, content) in file.lines
				),
			)
			u = self.outputdir/(filename + ".html")
			with contextlib.closing(u.openwrite()) as f:
				f.write(s.encode("utf-8"))

		# Copy CSS/JS/GIF files
		for filename in ("coverage.css", "coverage_sortfilelist.css", "coverage.js", "spc.gif"):
			self.logProgress("### copying %s" % filename)
			try:
				import pkg_resources
			except ImportError:
				data = open(os.path.join(os.path.dirname(__file__), filename), "rb").read()
			else:
				data = pkg_resources.resource_string(__name__, filename)
			with contextlib.closing((self.outputdir/filename).openwrite()) as f:
				f.write(data)

		self.logProgress("### creating buildlog.txt")
		with contextlib.closing((self.outputdir/"buildlog.txt").openwrite()) as f:
			f.write("".join(self.buildlog))

		self.logProgress("### creating testlog.txt")
		with contextlib.closing((self.outputdir/"testlog.txt").openwrite()) as f:
			f.write("".join(self.testlog))

	def execute(self):
		self.cleanup()
		self.download()
		self.unpack()
		self.configure()
		files = self.files("python")
		self.make()
		self.test()
		for file in files:
			if file.name.endswith(".py"):
				self.coverpy(file)
			elif file.name.endswith(".c"):
				self.coverc(file)
		self.makehtml(files)
		self.logLoop("done with project Python (%s; %d files)" % (self.timestamp.strftime("%Y-%m-%d %H:%M:%S"), len(files)))


def main(args=None):
	p = optparse.OptionParser(usage="usage: %prog [options]")
	p.add_option("-o", "--outputdir", dest="outputdir", help="Directory where to put the HTML files", default="~/pycoco")
	(options, args) = p.parse_args(args)
	if len(args) != 0:
		p.error("incorrect number of arguments")
		return 1
	sisyphus.execute(Python_GenerateCodeCoverage(options.outputdir))
	return 0


if __name__=="__main__":
	sys.exit(main())


src/pycoco.py
# -*- coding: iso-8859-1 -*-

import sys, os, datetime, urllib, optparse

try:
	import pkg_resources
except ImportError:
	css = open(os.path.join(os.path.dirname(__file__), "EGG-INFO", "coverage.css"), "rb").read()
else:
	css = pkg_resources.get_distribution("pycoco").get_metadata("coverage.css")


from ll import sisyphus, url
from ll.xist import xsc
from ll.xist.ns import xml, html, meta, htmlspecials


class xmlns(xsc.Namespace):
	xmlname = "cov"
	xmlurl = "http://xmlns.python.org/coverage"

	class page(xsc.Element):
		class Attrs(xsc.Element.Attrs):
			class title(xsc.TextAttr): required = True
			class crumbs(xsc.TextAttr): required = True

		def convert(self, converter):
			e = xsc.Frag(
				xml.XML10(), "\n",
				html.DocTypeXHTML10transitional(), "\n",
				html.html(
					html.head(
						meta.contenttype(),
						html.title(self.attrs.title),
						html.link(type="text/css", rel="stylesheet", href="root:coverage.css"),
					),
					html.body(
						html.div(
							html.div(
								html.a(
									htmlspecials.autoimg(src="http://www.python.org/images/python-logo.gif", alt="Python", border=0),
									href="http://www.python.org/",
								),
								class_="logo",
							),
							html.div(
								self.attrs.crumbs,
								class_="crumbs",
							),
							class_="header",
						),
						html.div(
							self.content,
							class_="content",
						),
					),
				),
			)
			return e.convert(converter)


	class crumb(xsc.Element):
		class Attrs(xsc.Element.Attrs):
			class href(xsc.URLAttr): pass
			class first(xsc.BoolAttr): pass

		def convert(self, converter):
			if self.attrs.first:
				c = u"\xbb"
			else:
				c = ">"
			e = self.content
			if self.attrs.href:
				e = html.a(e, href=self.attrs.href)
			else:
				e = html.span(e, class_="here")
			e = html.span(html.span(c, class_="bullet"), e, class_="crumb")
			return e.convert(converter)


	class filelist(xsc.Element):
		class Attrs(xsc.Element.Attrs):
			class timestamp(xsc.TextAttr): pass
			class revision(xsc.TextAttr): pass
		def convert(self, converter):
			now = datetime.datetime.now()
			e = xsc.Frag(
				html.h1("Python code coverage"),
				html.p("Generated at ", now.strftime("%Y-%m-%d %H:%M:%S"), class_="note"),
				html.p(self.attrs.timestamp, class_="note"),
				html.p(self.attrs.revision, class_="note"),
				htmlspecials.plaintable(
					html.thead(
						html.tr(
							html.th("Filename"),
							html.th("# lines"),
							html.th("# coverable lines"),
							html.th("# covered lines"),
							html.th("coverage"),
							html.th("distribution"),
						),
					),
					html.tbody(
						self.content,
					),
					class_="files",
				)
			)
			return e.convert(converter)


	class fileitem(xsc.Element):
		class Attrs(xsc.Element.Attrs):
			class name(xsc.TextAttr): required = True
			class lines(xsc.IntAttr): required = True
			class coverablelines(xsc.IntAttr): required = True
			class coveredlines(xsc.IntAttr): required = True

		def convert(self, converter):
			lines = int(self.attrs.lines)
			coverablelines = int(self.attrs.coverablelines)
			coveredlines = int(self.attrs.coveredlines)

			distsize = (100, 8)
			if coverablelines:
				coverage = "%.02f%%" % (100.*coveredlines/coverablelines)
				coverageclass = "int"
				distribution = xsc.Frag()
				totalwidth = 0
				if coverablelines < lines:
					width = int(float(lines-coverablelines)/lines*distsize[0])
					distribution.append(htmlspecials.pixel(width=width, height=distsize[1], style="background-color: #ccc;"))
					totalwidth += width
				if coveredlines < coverablelines:
					width = int(float(coverablelines-coveredlines)/lines*distsize[0])
					distribution.append(htmlspecials.pixel(width=width, height=distsize[1], style="background-color: #f00;"))
					totalwidth += width
				if totalwidth < distsize[0]:
					width = distsize[0]-totalwidth
					distribution.append(htmlspecials.pixel(width=width, height=distsize[1], style="background-color: #0c0;"))
					totalwidth += width
			else:
				coverage = "n/a"
				coverageclass = "int disable"
				distribution = htmlspecials.pixel(width=distsize[0], height=distsize[1], style="background-color: #000;")

			e = html.tr(
				html.th(
					html.a(
						self.attrs.name,
						href=("root:", self.attrs.name, ".html"),
					),
					class_="filename",
				),
				html.td(
					lines,
					class_="int",
				),
				html.td(
					coverablelines,
					class_="int",
				),
				html.td(
					coveredlines,
					class_="int",
				),
				html.td(
					coverage,
					class_=coverageclass,
				),
				html.td(
					distribution,
					class_="dist",
				),
				class_="files",
			)
			return e.convert(converter)


	class filecontent(xsc.Element):
		class Attrs(xsc.Element.Attrs):
			class name(xsc.TextAttr): required = True

		def convert(self, converter):
			e = xsc.Frag(
				html.h1("Python code coverage for ", self.attrs.name),
				htmlspecials.plaintable(
					html.thead(
						html.tr(
							html.th("#"),
							html.th("count"),
							html.th("content"),
						),
					),
					html.tbody(
						self.content,
					),
					class_="file",
				)
			)
			return e.convert(converter)


	class fileline(xsc.Element):
		class Attrs(xsc.Element.Attrs):
			class lineno(xsc.IntAttr): required = True
			class count(xsc.IntAttr): required = True

		def convert(self, converter):
			class_ = None
			count = None
			if self.attrs.count:
				count = int(self.attrs.count)
				if not count:
					class_ = "uncovered"
				elif count == -1:
					class_ = "uncoverable"
					count = "n/a"
			else:
				class_ = "uncoverable"
				count = "n/a"
			e = html.tr(
				html.th(self.attrs.lineno),
				html.td(count, class_="count"),
				html.td(self.content, class_="line"),
				class_=class_
			)
			return e.convert(converter)


class File(object):
	def __init__(self, name):
		self.name = name
		self.lines = [] # list of lines with tuples (# of executions, line)

	def __repr__(self):
		return "<File name=%r at 0x%x>" % (self.name, id(self))


class Python_GenerateCodeCoverage(sisyphus.Job):
	def __init__(self, outputdir):
		sisyphus.Job.__init__(self, 60*60, name="Python_GenerateCodeCoverage", raiseerrors=1)
		self.url = "http://svn.python.org/snapshots/python.tar.bz2"
		self.tarfile = "python.tar.bz2"
		self.outputdir = url.Dir(outputdir)

		self.configurecmd = "./configure --enable-unicode=ucs4 --with-pydebug"
		self.compileopts = "-fprofile-arcs -ftest-coverage"
		self.linkopts = "-lgcov"
		self.gcovcmd = os.environ.get("COV", "gcov")
		self.makefile = "python/Makefile"

		self.buildlog = [] # the output of configuring and building Python
		self.testlog = [] # the output of running the test suite

	def cmd(self, cmd):
		self.logProgress(">>> %s" % cmd)
		pipe = os.popen(cmd + " 2>&1")
		lines = []
		for line in pipe:
			self.logProgress("... " + line)
			lines.append(line)
		return lines

	def files(self, base):
		self.logProgress("### finding files")
		allfiles = []
		for (root, dirs, files) in os.walk(base):
			for file in files:
				if file.endswith(".py") or file.endswith(".c"):
					allfiles.append(File(os.path.join(root, file)))
		self.logProgress("### found %d files" % len(allfiles))
		return allfiles

	def download(self):
		self.logProgress("### downloading %s to %s" % (self.url, self.tarfile))
		urllib.urlretrieve(self.url, self.tarfile)

	def unpack(self):
		self.logProgress("### unpacking %s" % self.tarfile)
		self.cmd("tar xvjf %s" % self.tarfile)
		lines = list(open("python/.timestamp", "r"))
		self.timestamp = datetime.datetime.fromtimestamp(int(lines[0]))
		self.revision = lines[2]

	def configure(self):
		self.logProgress("### configuring")
		lines = self.cmd("cd python; %s" % self.configurecmd)
		self.buildlog.extend(lines)
		makelines = []
		self.logProgress("### adding compiler options %s" % self.compileopts)
		for line in open(self.makefile, "r"):
			if line.startswith("OPT") and line[3:].strip().startswith("="):
				line = line.rstrip("\n") + " " + self.compileopts + "\n"
			if line.startswith("LIBC") and line[4:].strip().startswith("="):
				line = line.rstrip("\n") + " " + self.linkopts + "\n"
			makelines.append(line)
		file = open(self.makefile, "w")
		file.writelines(makelines)
		file.close()

	def make(self):
		self.logProgress("### running make")
		self.buildlog.extend(self.cmd("cd python && make"))

	def test(self):
		self.logProgress("### running test")
		lines = self.cmd("cd python && ./python Lib/test/regrtest.py -T -N -R :: -uurlfetch,largefile,network,decimal")
		self.testlog.extend(lines)

	def cleanup(self):
		self.logProgress("### cleaning up files from previous run")
		self.cmd("rm -rf python")
		self.cmd("rm %s" % self.tarfile)

	def coveruncovered(self, file):
		self.logProgress("### faking coverage info for uncovered file %r" % file.name)
		file.lines = [(None, line.rstrip("\n")) for line in open(file.name, "r")]

	def coverpy(self, file):
		coverfilename = os.path.splitext(file.name)[0] + ".cover"
		self.logProgress("### fetching coverage info for Python file %r from %r" % (file.name, coverfilename))
		try:
			f = open(coverfilename, "r")
		except IOError, exc:
			self.logError(exc)
			self.coveruncovered(file)
		else:
			for line in f:
				line = line.rstrip("\n")
				prefix, line = line[:7], line[7:]
				prefix = prefix.strip()
				if prefix == "." or prefix == "":
					file.lines.append((-1, line))
				elif prefix == ">"*len(prefix):
					file.lines.append((0, line))
				else:
					file.lines.append((int(prefix.rstrip(":")), line))
			f.close()

	def coverc(self, file):
		self.logProgress("### fetching coverage info for C file %r" % file.name)
		dirname = os.path.dirname(file.name).split("/", 1)[-1]
		basename = os.path.basename(file.name)
		self.cmd("cd python && %s %s -o %s" % (self.gcovcmd, basename, dirname))
		try:
			f = open("python/%s.gcov" % basename, "r")
		except IOError, exc:
			self.logError(exc)
			self.coveruncovered(file)
		else:
			for line in f:
				line = line.rstrip("\n")
				if line.count(":") < 2:
					continue
				(count, lineno, line) = line.split(":", 2)
				count = count.strip()
				lineno = lineno.strip()
				if lineno == "0": # just the header
					continue
				if count == "-": # not executable
					file.lines.append((-1, line))
				elif count == "#####": # not executed
					file.lines.append((0, line))
				else:
					file.lines.append((int(count), line))
			f.close()

	def makehtml(self, files):
		ns = xmlns
		# Generate page for each source file
		for (i, file) in enumerate(files):
			filename = file.name.split("/", 1)[-1]
			self.logProgress("### generating HTML %d/%d for %s" % (i+1, len(files), filename))
			e = ns.page(
				ns.filecontent(name=filename)(
					ns.fileline(
						content.decode("latin-1").expandtabs(8),
						lineno=i+1,
						count=count,
					)
					for (i, (count, content)) in enumerate(file.lines),
				),
				title=("Python code coverage: ", filename),
				crumbs=(
					ns.crumb("Core Development", href="http://www.python.org/dev/", first=True),
					ns.crumb("Code coverage", href="root:index.html"),
					ns.crumb(filename),
				),
			)
			e = e.conv()
			u = self.outputdir/(filename + ".html")
			e.write(u.openwrite(), base="root:%s.html" % filename, encoding="utf-8")

		# Generate main page
		self.logProgress("### generating index page")
		e = ns.page(
			ns.filelist(
				(
					ns.fileitem(
						name=file.name.split("/", 1)[-1],
						lines=len(file.lines),
						coverablelines=sum(line[0]>=0 for line in file.lines),
						coveredlines=sum(line[0]>0 for line in file.lines),
					)
					for file in files
				),
				timestamp=("Repository timestamp ", self.timestamp.strftime("%Y-%m-%d %H:%M:%S")),
				revision=self.revision,
			),
			title=("Python code coverage (", self.timestamp.strftime("%Y-%m-%d"), ")"),
			crumbs=(
				ns.crumb("Core Development", href="http://www.python.org/dev/", first=True),
				ns.crumb("Code coverage"),
			),
		)
		e = e.conv()
		u = self.outputdir/"index.html"
		e.write(u.openwrite(), base="root:index.html", encoding="utf-8")

		# Copy CSS file
		self.logProgress("### saving CSS file")
		(self.outputdir/"coverage.css").openwrite().write(css)

	def execute(self):
		self.cleanup()
		self.download()
		self.unpack()
		self.configure()
		files = self.files("python")
		self.make()
		self.test()
		for file in files:
			if file.name.endswith(".py"):
				self.coverpy(file)
			elif file.name.endswith(".c"):
				self.coverc(file)
		self.makehtml(files)
		self.logLoop("done with project Python (%s; %d files)" % (self.timestamp.strftime("%Y-%m-%d %H:%M:%S"), len(files)))


def main(args=None):
	p = optparse.OptionParser(usage="usage: %prog [options]")
	p.add_option("-o", "--outputdir", dest="outputdir", help="Directory where to put the HTML files", default="~/pycoco")
	(options, args) = p.parse_args(args)
	if len(args) != 0:
		p.error("incorrect number of arguments")
		return 1
	sisyphus.execute(Python_GenerateCodeCoverage(options.outputdir))
	return 0


if __name__=="__main__":
	sys.exit(main())
